lib/profileAgents.js

```diff
@@ -0,0 +1,319 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+require('bedrock-account');
+const assert = require('assert-plus');
+const base64url = require('base64url-universal');
+const bedrock = require('bedrock');
+const crypto = require('crypto');
+const {CapabilityAgent} = require('webkms-client');
+const database = require('bedrock-mongodb');
+const {promisify} = require('util');
+const kms = require('./kms');
+const utils = require('./utils');
+const zcaps = require('./zcaps');
+
+const {util: {BedrockError}} = bedrock;
+
+// load config defaults
+require('./config');
+const KMS_MODULE = 'ssm-v1';
+
+// const PERMISSIONS = bedrock.config.permission.permissions;
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await promisify(database.openCollections)(['profile-profileAgent']);
+
+  await promisify(database.createIndexes)([{
+    // cover queries of a profile agent by its ID
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.id': 1},
+    options: {unique: true, background: false}
+  }, {
+    // cover queries of a profile agent by its ID
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.profile': 1},
+    options: {unique: false, background: false}
+  }, {
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.controller': 1},
+    options: {unique: true, background: false}
+  }, {
+    // cover queries of a profile agent by an account ID
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.account': 1, 'profileAgent.id': 1},
+    options: {
+      partialFilterExpression: {account: {$exists: true}},
+      unique: true,
+      background: false
+    }
+  }]);
+});
+
+/**
+ * Creates a ProfileAgent and associates it to an account if provided
+ *
+ * @param {Object} account the account creating the ProfileAgent.
+ *
+ * @return {Promise<ProfileAgent>} resolves to a ProfileAgent.
+ */
+exports.create = async ({account}) => {
+  // 1. Generate a random secret
+  const secret = base64url.encode(crypto.randomBytes(32));
+  const handle = 'primary';
+  // 2. Generate capability agent for the zCap key
+  const capabilityAgent = await CapabilityAgent.fromSecret({handle, secret});
+  // 3. Create keystore in order to create the zCap key for the profile agent
```

Comment:
We might want to store the capability agent up here first, otherwise we have no clean way of cleaning up the other stuff that gets created on other systems should we experience a failure here.

Reply:
Added a TODO for this and will punt for now.